### PR TITLE
Bug/#7417 - Update Module to `analytics-4` for GA4 Dashboard View Auto-Switching

### PR DESCRIPTION
**Pull request by hussain-t**
## Summary

Addresses issue:

- #7417 

## PR Author Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [ ] My code is backward-compatible with WordPress 5.2 and PHP 5.6.
- [ ] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [ ] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [x] Run the code.
- [x] Ensure the acceptance criteria are satisfied.
- [x] Reassess the implementation with the IB.
- [x] Ensure no unrelated changes are included.
- [x] Ensure CI checks pass.
- [ ] Check Storybook where applicable.
- [x] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [x] Ensure the PR has the correct target branch.
- [x] Double-check that the PR is okay to be merged.
- [x] Ensure the corresponding issue has a ZenHub release assigned.
- [x] Add a changelog message to the issue.

### CODE REVIEW

File: assets/js/googlesitekit/datastore/user/permissions.test.js [hunk 1]
```diff
@@ -481,6 +565,50 @@ describe( 'core/user authentication', () => {
 
 				expect( canViewSharedModule ).toBe( true );
 			} );
+
+			it( 'should change module slug to "analytics-4" when the `module.slug` is "analytics" and the dashboard view is GA4', () => {
```
Comment (techanvil) [inline assets/js/googlesitekit/datastore/user/permissions.test.js:569]: This test is good, but I'd suggest rewording the description a bit to describe it more in terms of the outcome as opposed to the underlying implementation:

```suggestion
			it( 'should treat `analytics` as `analytics-4` when the dashboard view is GA4', () => {
```

File: assets/js/googlesitekit/datastore/user/permissions.test.js [hunk 2]
```diff
@@ -481,6 +565,50 @@ describe( 'core/user authentication', () => {
...
+				enabledFeatures.add( 'ga4Reporting' );
+
+				registry.dispatch( CORE_USER ).receiveGetCapabilities( {
+					...capabilitiesWithPermission.permissions,
+					// Set `analytics` permission to `false` to ensure that the module slug is changed to "analytics-4".
+					// The selector will return `true` as the module slug is changed to "analytics-4".
```
Comment (techanvil) [inline assets/js/googlesitekit/datastore/user/permissions.test.js:575]: As with the description, I'd suggest a tweak here to make it _slightly_ less implementation-specific:

```suggestion
					// Set the `analytics` permission to `false` to help verify that the `analytics` module is indeed treated as `analytics-4` when the dashboard view is GA4.
```

File: assets/js/googlesitekit/datastore/user/permissions.test.js [hunk 3]
```diff
@@ -361,8 +384,69 @@ describe( 'core/user authentication', () => {
 					'search-console',
 					'analytics',
 					'pagespeed-insights',
+				] );
+			} );
+
+			it( 'should not include "analytics" module if the dashboardView is GA4', () => {
```
Comment (techanvil) [inline assets/js/googlesitekit/datastore/user/permissions.test.js:390]: Can we change the quotes in these descriptions just to make the formatting consistent throughout these new descriptions and comments? Please can you apply this throughout the PR :)

```suggestion
			it( 'should not include `analytics` module if the dashboardView is GA4', () => {
```

File: assets/js/googlesitekit/datastore/user/permissions.test.js [hunk 4]
```diff
@@ -481,6 +565,50 @@ describe( 'core/user authentication', () => {
...
+					},
+				] );
+
+				const canViewSharedAnalytics = registry
+					.select( CORE_USER )
+					.canViewSharedModule( 'analytics' );
+
+				expect( canViewSharedAnalytics ).toBe( false );
+
+				// Set the dashboard view to GA4 to change the module slug to "analytics-4".
```
Comment (techanvil) [inline assets/js/googlesitekit/datastore/user/permissions.test.js:599]: I think we can then simply remove this comment...

```suggestion
```

File: assets/js/googlesitekit/datastore/user/permissions.js [hunk 5]
```diff
@@ -239,6 +239,14 @@ const baseSelectors = {
 		// shareable and the user has the "read shared module data"
 		// capability for.
 		return Object.values( modules ).reduce( ( slugs, module ) => {
+			if ( module.slug === 'analytics' && isGA4DashboardView ) {
+				return slugs;
+			}
+
+			if ( module.slug === 'analytics-4' && ! isGA4DashboardView ) {
+				return slugs;
+			}
```
Comment (techanvil) [inline assets/js/googlesitekit/datastore/user/permissions.js:248]: I know this is how I formatted the code in my PoC, but I would really prefer to see these conditions combined (or at least use an `else` to make it clear from a quick scan that the conditions are mutually exclusive). What do you think?

```suggestion
			if (
				( module.slug === 'analytics' && isGA4DashboardView ) ||
				( module.slug === 'analytics-4' && ! isGA4DashboardView )
			) {
				return slugs;
			}
```

File: assets/js/googlesitekit/modules/datastore/modules.js [hunk 6]
```diff
@@ -1393,7 +1393,18 @@ const baseSelectors = {
...
+			select( MODULES_ANALYTICS ).isGA4DashboardView();
+
 		return Object.keys( modules ).reduce( ( acc, slug ) => {
+			if ( slug === 'analytics' && isGA4DashboardView ) {
+				return acc;
+			}
+
+			if ( slug === 'analytics-4' && ! isGA4DashboardView ) {
+				return acc;
+			}
```
Comment (techanvil) [inline assets/js/googlesitekit/modules/datastore/modules.js:1406]: Similarly to my other comment, I think these would be better off with the conditions combined:

```suggestion
			if (
				( slug === 'analytics' && isGA4DashboardView ) ||
				( slug === 'analytics-4' && ! isGA4DashboardView )
			) {
				return acc;
			}
```

**Review by techanvil — CHANGES_REQUESTED** [review]
Hey @hussain-t, nice work here. I have left a few comments - they are all very minor, apologies for the finicky nature of some of them ;)

File: assets/js/googlesitekit/datastore/user/permissions.js (hunk 5 above)
Comment (hussain-t) [reply assets/js/googlesitekit/datastore/user/permissions.js:248]: SGTM 👍

File: assets/js/googlesitekit/modules/datastore/modules.js (hunk 6 above)
Comment (hussain-t) [reply assets/js/googlesitekit/modules/datastore/modules.js]: Updated ✅

File: assets/js/googlesitekit/datastore/user/permissions.js (hunk 5 above)
Comment (hussain-t) [reply assets/js/googlesitekit/datastore/user/permissions.js]: Updated ✅

File: assets/js/googlesitekit/datastore/user/permissions.test.js (hunk 1 above)
Comment (hussain-t) [reply assets/js/googlesitekit/datastore/user/permissions.test.js]: Updated ✅

File: assets/js/googlesitekit/datastore/user/permissions.test.js (hunk 3 above)
Comment (techanvil) [reply assets/js/googlesitekit/datastore/user/permissions.test.js]: Thanks for applying this change. I noticed there were still a couple of cases in this format in `modules.test.js`, so I've pushed an update for this myself to save the round trip :)

**Review by techanvil — APPROVED** [review]
LGTM, nice one @hussain-t!
